### PR TITLE
Allow listing firmware versions

### PR DESCRIPTION
**Pull request by jbliesener**
This change allows fwupdate to list available firmware versions. `${OS_FIRWARE_VERSIONS_STABLE}` and `${OS_FIRWARE_VERSIONS_BETA}` contain two URLs that may contain the usual placeholders (`\${platform}`,`\${os_prefix}`, `\${os_short_name}` and `\${version}`) and each download a JSON file that contains a list of available stable or beta versions.

Each of the files contains a list (array) of JSON objects with the following fields:

- `version`: the version number
- `date`: the release date
- `url`: the url to download the version

The output of the command is a table sorted in reverse semver-order that, for each found version, contains a line with the three fields mentioned above

Example:

Platform is `raspberrypi`

```
OS_PRERELEASES="true"
OS_FIRMWARE_VERSIONS_STABLE="https://www.example.com/releases/latest/downloads/versions_release_\${platform}.json"
OS_FIRMWARE_VERSIONS_BETA="https://www.example.com/releases/latest/downloads/versions_prerelease_\${platform}.json"
```

File `versions_release_raspberrypi.json`, can be reached at https://www.example.com/releases/latest/downloads/versions_release_raspberrypi.json:

```
[
  {
    "url": "https://www.example.com/releases/0.5.2/downloads/thingos-raspberrypi-0.5.2.img.xz",
    "version": "0.5.2",
    "date": "2022-12-19"
  }
]
```

File `versions_prerelease_raspberrypi.json`, can be reached at https://www.example.com/releases/latest/downloads/versions_prerelease_raspberrypi.json:
```
[
  {
    "url": "https://www.example.com/releases/0.5.2-beta/downloads/thingos-raspberrypi-0.5.2-beta.img.xz",
    "version": "0.5.2-beta",
    "date": "2022-12-17"
  },
  {
    "url": "https://www.example.com/releases/0.5.3-snapshot/downloads/thingos-raspberrypi-0.5.3-snapshot.img.xz",
    "version": "0.5.3-snapshot",
    "date": "2023-04-18"
  }
]
```

Output of `fwupdate list` or `fwupdate list all`:
```
0.5.3-snapshot  2023-04-18      https://www.example.com/releases/0.5.3-snapshot/downloads/thingos-raspberrypi-0.5.3-snapshot.img.xz
0.5.2   2022-12-19      https://www.example.com/releases/0.5.2/downloads/thingos-raspberrypi-0.5.2.img.xz
0.5.2-beta      2022-12-17      https://www.example.com/releases/0.5.2-beta/downloads/thingos-raspberrypi-0.5.2-beta.img.xz
```

Output of `fwupdate list stable`:
```
0.5.2   2022-12-19      https://www.example.com/releases/0.5.2/downloads/thingos-raspberrypi-0.5.2.img.xz
```

Output of `fwupdate list beta`:
```
0.5.3-snapshot  2023-04-18      https://www.example.com/releases/0.5.3-snapshot/downloads/thingos-raspberrypi-0.5.3-snapshot.img.xz
0.5.2-beta      2022-12-17      https://www.example.com/releases/0.5.2-beta/downloads/thingos-raspberrypi-0.5.2-beta.img.xz
```

If `${OS_PRERELEASES}=="false"`, `fwupdate list` without any additional parameter produces the same result as `fwupdate list stable`

### CODE REVIEW

File: board/common/overlay/sbin/fwupdate [hunk 1]
```diff
@@ -37,11 +41,6 @@ function exit_usage() {
     exit 1
 }
 
-if [[ -z "$1" ]]; then
```
Comment (ccrisan) [inline board/common/overlay/sbin/fwupdate:40]: Why is this check gone? Do we ever want to run `fwupdate` without a subcommand and not show the usage?

Comment (jbliesener) [reply board/common/overlay/sbin/fwupdate:40]: It's not gone, it's just executed further down, when variables are available. These variables are required in exit_usage to decide whether the `fwupdate list` feature will be shown.

File: board/common/overlay/sbin/fwupdate [hunk 2]
```diff
@@ -271,6 +273,38 @@ function show_latest() {
 }
 
 
+#### Show all versions ####
+
+function list_versions() {
+    check_versions=${1,,}
+    if [[ -z $check_versions ]]; then
```
[the remark this answers is not in the document]
Comment (jbliesener) [reply board/common/overlay/sbin/fwupdate:280]: Right, sorry for that. Will fix.

File: board/common/overlay/sbin/fwupdate [hunk 3]
```diff
@@ -8,6 +8,10 @@ function exit_usage() {
     echo "           shows the current version"
     echo "       fwupdate latest"
     echo "           shows the latest available version with details"
+if [[ -n ${OS_FIRMWARE_VERSIONS_STABLE} && -n {OS_FIRMWARE_VERSIONS_BETA} ]]; then
```
Comment (ccrisan) [inline board/common/overlay/sbin/fwupdate:11]: For the sake of consistency, i'd prefer to always have this option displayed (just like for the "latest" option), even when corresponding variables are not set.

I think it's better to have a simple, unified help text and to do the check in the corresponding subcommand function. I know this doesn't currently happen if `OS_FIRMWARE_LATEST_STABLE` is, for example, unset.

File: board/common/overlay/sbin/fwupdate (hunk 1 above)
Comment (ccrisan) [reply board/common/overlay/sbin/fwupdate:40]: Oh, I somehow missed it.

**Comment by ccrisan** [comment]
Nice addition to the fwupdate command!

**Comment by jbliesener** [comment]
Changes implemented as discussed

**Comment by ccrisan** [comment]
I've also updated the Wiki page accordingly: https://github.com/ccrisan/thingos/wiki/Firmware-Updates#versions.

**Pull request merged**